From da706d429a949b13f10cf3651495fb2436109253 Mon Sep 17 00:00:00 2001
From: Penny Zheng <Penny.Zheng@arm.com>
Date: Mon, 9 Oct 2023 12:53:44 +0800
Subject: mbedtls: Fix dependencies with PSA crypto enabled

After an update to mbed TLS 3.3.0, dependencies with
CONFIG_MBEDTLS_PSA_CRYPTO_C enabled got affected.

mbed TLS in its build_info.h file, enables MBEDTLS_PK_WRITE_C config
under the hood. MBEDTLS_PK_WRITE_C has a dependency to
MBEDTLS_ASN1_WRITE_C which wasn't reflected anywhere.

Therefore, update Kconfig.tls-generic to enable
CONFIG_MBEDTLS_PK_WRITE_C automatically, when PSA crypto is enabled, to
reflect mbed TLS configuration pattern. Additionally, enable
MBEDTLS_ASN1_WRITE_C automatically, when PK write is enabled.

Upstream-Status: Backport
Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>
---
 modules/mbedtls/CMakeLists.txt               |  2 +-
 modules/mbedtls/Kconfig.tls-generic          |  3 ++-
 modules/mbedtls/configs/config-tls-generic.h | 17 ++++++++---------
 3 files changed, 11 insertions(+), 11 deletions(-)

diff --git a/modules/mbedtls/CMakeLists.txt b/modules/mbedtls/CMakeLists.txt
index 09017a5b83..aee2f78bd1 100644
--- a/modules/mbedtls/CMakeLists.txt
+++ b/modules/mbedtls/CMakeLists.txt
@@ -107,7 +107,7 @@ zephyr_interface_library_named(mbedTLS)
 
     zephyr_library_named(mbedTLSCrypto)
 
-    if (MBEDTLS_USE_PSA_CRYPTO)
+    if (CONFIG_MBEDTLS_PSA_CRYPTO_C)
       list(APPEND crypto_source
         ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_aead.c
         ${ZEPHYR_CURRENT_MODULE_DIR}/library/psa_crypto_cipher.c
diff --git a/modules/mbedtls/Kconfig.tls-generic b/modules/mbedtls/Kconfig.tls-generic
index 6275c6570f..a891759b40 100644
--- a/modules/mbedtls/Kconfig.tls-generic
+++ b/modules/mbedtls/Kconfig.tls-generic
@@ -413,7 +413,8 @@ config MBEDTLS_SERVER_NAME_INDICATION
 
 config MBEDTLS_PK_WRITE_C
 	bool "The generic public (asymmetric) key writer"
-	help
+	default y if MBEDTLS_PSA_CRYPTO_C
+ 	help
 	  Enable generic public key write functions.
 
 config MBEDTLS_HAVE_TIME_DATE
diff --git a/modules/mbedtls/configs/config-tls-generic.h b/modules/mbedtls/configs/config-tls-generic.h
index b95d0f1f0f..fd5cc83289 100644
--- a/modules/mbedtls/configs/config-tls-generic.h
+++ b/modules/mbedtls/configs/config-tls-generic.h
@@ -390,15 +390,6 @@
 #define MBEDTLS_X509_USE_C
 #endif
 
-#if defined(MBEDTLS_X509_USE_C) || \
-    defined(MBEDTLS_ECDSA_C)
-#define MBEDTLS_ASN1_PARSE_C
-#endif
-
-#if defined(MBEDTLS_ECDSA_C)
-#define MBEDTLS_ASN1_WRITE_C
-#endif
-
 #if defined(MBEDTLS_DHM_C) || \
     defined(MBEDTLS_ECP_C) || \
     defined(MBEDTLS_RSA_C) || \
@@ -424,6 +415,14 @@
 #define MBEDTLS_PK_C
 #endif
 
+#if defined(MBEDTLS_X509_USE_C) || defined(MBEDTLS_ECDSA_C)
+#define MBEDTLS_ASN1_PARSE_C
+#endif
+
+#if defined(MBEDTLS_ECDSA_C) || defined(MBEDTLS_PK_WRITE_C)
+#define MBEDTLS_ASN1_WRITE_C
+#endif
+
 #if defined(CONFIG_MBEDTLS_PKCS5_C)
 #define MBEDTLS_PKCS5_C
 #endif
-- 
2.25.1

